cmd/zrepl/mgmt_conn.c

```diff
@@ -538,6 +560,115 @@ uzfs_zvol_rebuild_status(uzfs_mgmt_conn_t *conn, zvol_io_hdr_t *hdrp,
 	return (reply_data(conn, &hdr, &status_ack, sizeof (status_ack)));
 }
 
+/*
+ * Iterate through all finished async tasks and send replies to clients.
+ */
+static int
+finish_async_tasks(void)
+{
+	uzfs_mgmt_conn_t *conn;
+	async_arg_t *async_task;
+	int rc;
+
+	mutex_enter(&async_tasks_mtx);
+	while ((async_task = SLIST_FIRST(&async_tasks)) != NULL) {
+		SLIST_REMOVE_HEAD(&async_tasks, task_next);
+		/* connection could have been removed in the meantime */
+		mutex_enter(&conn_list_mtx);
+		SLIST_FOREACH(conn, &uzfs_mgmt_conns, conn_next) {
+			if (conn == async_task->ctx)
+				break;
+		}
+		mutex_exit(&conn_list_mtx);
```

Comment:
after this 'mutex_exit', conn data structure might have got deleted right..?

Reply:
yes, it is unlikely but IMO possible. The connection is not destroyed unless the zvol is destroyed. Once the async op finishes and drops reference on zvol, it can be in theory destroyed. So either before or after we grab the mutex here, the conn can be destroyed.

Reply:
how about doing 'reply_nodata' also inside the lock 'conn_list_mtx'?
Also, we need to check in reply_nodata whether conn_fd is -1 or not before doing epoll_ctl..

Reply:
we cannot do reply_nodata from here because sockets and state structures can only be touched from main event loop thead. And for the same reason checking conn_fd == -1 is useless in reply_nodata. If conn_fd was closed and set to -1 then reply_nodata would not be called - it's all done by a single thread.

Reply:
my mistake, this is not executed on behalf of async thread but on behalf of main event loop thread. So we could do reply_nodata with held mutex but it doesn't help. conn_fd cannot change in the meantime anyway.

Reply:
(after a short chat with Vitta on Slack and long thinking on my side) ... Vitta is right! :-) There is indeed a problem in this piece of code. Consider following scenario:

1) async task is dispatched and runs in its own thread
2) async task finishes and drops ref count on dataset.
3) async tasks puts the finished task to queue and notifies main event loop. <-- RACE BEGINS
4) dataset is destroyed. conn structure continues to exist but it is marked as unused (refcount = 0). event loop thread is notified.
5) event loop thread gets notified and closes the connection and sets FD to -1.
6) event loop  thread gets notified and finds the finished async task. conn struct is still there, but FD has already been set to -1. It tries to send a reply for async task on this connection --> PROBLEM

This is a flaw introduced on behalf of async tasks to mgmt_conn.c. Thanks Vitta (and Satbir) for spotting it! Let's think about solution. Requirements for the solution:

1) zfs destroy should not block because of pending async reply
2) connection should be closed after async reply is sent (no new commands)
3) if new connection to the same iscsi target is demanded (by a new zvol) and the connection has not been destroyed because of pending async task, we can reuse it without closing it.

I suggest solution when we grab reference on conn every time we dispatch new async task. That will guarantee the conn struct to stay until the async command is finished. zvol can be destroyed in the meantime and in that case the conn will be destroyed too after async command has finished and reply has been sent. We will need a new connection state to implement this and make a couple of changes. If you agree with this approach in general, I will submit the new changes for review.